hack/lib/util/sync.sh

```diff
@@ -0,0 +1,110 @@
+#!/bin/bash
+#
+# This library holds sync utility functions.
+
+# os::sync::fork is responsible for synchronizing specified fork from UPSTREAM commits
+# in our vendor directory
+#
+# Arguments:
+#  - 1 - temporary directory where synchronization will take place, if not passed one will be created for you
+#  - 2 - the source repository from which take the upstream cherry-picks
+#  - 3 - the branch in source repository
+#  - 4 - the directory inside the source repository where changes are applied
+#  - 5 - the destination repository (fork) where changes are targeted
+#  - 6 - the destination repository branch
+# Returns:
+#  None
+function os::sync::fork() {
+    local tmpDir="$1"
+    local fromRepo="$2"
+    local fromBranch="$3"
+    local fromDir="$4"
+    local toRepo="$5"
+    local toBranch="$6"
+    local fromRepoName=$(basename ${fromRepo})
+    local toRepoName=$(basename ${toRepo})
+
+    fromRepoLocation="${tmpDir}/${fromRepoName}"
+    toRepoLocation="${tmpDir}/${toRepoName}"
+
+    mkdir -p "${fromRepoLocation}"
+    mkdir -p "${toRepoLocation}"
+
+    pushd "${fromRepoLocation}" > /dev/null
+    # if we already have a git repo, assume it's the right one and get the requested branch
+    if ! git rev-parse --quiet --is-inside-work-tree &> /dev/null; then
+        git clone --quiet --single-branch --branch="${fromBranch}" "${fromRepo}" "${fromRepoLocation}"
+    fi
+    git fetch --refmap -- origin refs/heads/"${fromBranch}":refs/remotes/origin/"${fromBranch}"
+    git checkout --quiet "${fromBranch}"
+    git reset --quiet --hard origin/"${fromBranch}"
+    firstFromSHA=$(git rev-list --max-parents=0 HEAD)
+    newFromSHA=$(git log --oneline --format='%H' -1)
+    popd > /dev/null
+
+    pushd "${toRepoLocation}" > /dev/null
+    # if we aren't in a git repo, clone it
+    if ! git rev-parse --quiet --is-inside-work-tree &> /dev/null; then
+        git clone --quiet --single-branch --branch="${toBranch}" "${toRepo}" "${toRepoLocation}"
+    fi
+    git fetch --refmap -- origin refs/heads/"${toBranch}":refs/remotes/origin/"${toBranch}"
+    git checkout --quiet "${toBranch}"
+    git reset --quiet --hard origin/"${toBranch}"
+    startingFromSHA="${firstFromSHA}"
+    if [ -f "${fromRepoName}.sha" ]; then
+        startingFromSHA=$(cat "${fromRepoName}.sha")
+    fi
+    if [ "${newFromSHA}" == "${startingFromSHA}" ]; then
+        popd > /dev/null
+        echo "already at level: ${newFromSHA}, nothing to sync"
+        exit 0
+    fi
+
+    os::log::info ""
+    os::log::info "Syncing from ${startingFromSHA}..${newFromSHA}"
+
+    # create a clean branch to start from
```

Comment:
There's a lot of logic here for cleaning, assuming we already cloned, etc ... I'd get rid of it all